### PR TITLE
dac: max blob size

### LINKED ISSUE #1233: Impose blob size limit in the manager before calling DALC SubmitBlocks
Currently all blocks from the pendingBlocks queue is pulled to call `SubmitBlocks` using DALC, which may lead to DA submission failures if the total blob size exceeds the limit. May be DALC interface provides a method to get the blob size limit which can then be used by the caller (manager).

### PR DESCRIPTION
**Pull request by tuxcanfly**
## Overview

Fixes #1233 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

Depends on https://github.com/rollkit/go-da/pull/23

## Summary by CodeRabbit

- **New Features**
  - Improved block submission process with retries and exponential backoff for increased reliability.
  - Added error handling for block size limitations and existence issues during submission.

- **Enhancements**
  - Block submission now includes a count of successfully submitted blocks for better tracking.
  - Partial reset functionality added to the management of pending blocks based on submission status.

- **Bug Fixes**
  - Fixed an issue where block submission could fail silently by ensuring all pending blocks are confirmed as submitted before completion.

- **Documentation**
  - Updated function documentation to reflect new error handling and submission count features.

### CODE REVIEW

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]

File: block/manager.go [hunk 1]
```diff
@@ -712,21 +712,24 @@
...
 	for attempt := 1; ctx.Err() == nil && !submitted && attempt <= maxSubmitAttempts; attempt++ {
-		res := m.dalc.SubmitBlocks(ctx, m.pendingBlocks.getPendingBlocks())
+		blocks := m.pendingBlocks.getPendingBlocks()
+		res := m.dalc.SubmitBlocks(ctx, blocks)
 		if res.Code == da.StatusSuccess {
-			m.logger.Info("successfully submitted Rollkit block to DA layer", "daHeight", res.DAHeight)
-			submitted = true
+			m.logger.Info("successfully submitted Rollkit block to DA layer", "daHeight", res.DAHeight, "count", res.Count)
+			if int(res.Count) == len(blocks) {
+				submitted = true
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply block/manager.go:720]: Thanks for clarifying

File: da/da.go [hunk 2]
```diff
@@ -39,6 +39,8 @@ type BaseResult struct {
 	Message string
 	// DAHeight informs about a height on Data Availability Layer for given result.
 	DAHeight uint64
+	// Count returns number of successfully submitted blocks.
+	Count uint64
```
[the remark this answers is not in the document]
Comment (tuxcanfly) [reply da/da.go]: Done, renamed to `SubmittedCount`

**Review by Manav-Aggarwal — DISMISSED** [review]
LGTM

File: da/da.go [hunk 3]
```diff
@@ -65,7 +72,18 @@ type DAClient struct {
...
 func (dac *DAClient) SubmitBlocks(ctx context.Context, blocks []*types.Block) ResultSubmitBlocks {
-	blobs := make([][]byte, len(blocks))
+	var blobs [][]byte
+	var blobSize uint64
+	maxBlobSize, err := dac.DA.MaxBlobSize()
+	if err != nil {
+		return ResultSubmitBlocks{
+			BaseResult: BaseResult{
+				Code:    StatusError,
+				Message: "unable to configure max blob size",
```
Comment (tzdybal) [inline da/da.go:82]:
```suggestion
				Message: "unable to get DA max blob size",
```

File: da/da_test.go [hunk 4]
```diff
@@ -37,6 +40,51 @@ func TestMain(m *testing.M) {
 	os.Exit(exitCode)
 }
 
+// MockDA is a mock for the DA interface
+type MockDA struct {
```
Comment (tzdybal) [inline da/da_test.go:44]: Why not Dummy from go-da?

**Review by MSevey — APPROVED** [review]

**Pull request merged**